src/szk.rs

```diff
@@ -0,0 +1,444 @@
+                ..
+            } => Cow::Owned(self.derive_helper_proofs_share(proof_share_seed_and_blind)),
+        };
+
+        let (joint_rand, jr_seed_and_part) = if self.typ.joint_rand_len() > 0 {
+            let (seed, other_joint_rand_part) = match proof {
+                SzkProofShare::Leader {
+                    uncompressed_proof_share: _,
+                    blind_and_joint_rand_part,
+                } => blind_and_joint_rand_part.expect("jr_len > 0 so there is joint_randomness"),
+                SzkProofShare::Helper {
+                    proof_share_seed_and_blind,
+                    joint_rand_part,
+                } => (
+                    proof_share_seed_and_blind,
+                    joint_rand_part.expect("jr_len > 0 so there is joint randomness"),
+                ),
+            };
+            let own_joint_rand_part = self.derive_joint_rand_part(&seed, input_share, nonce)?;
+            let (leader_joint_rand_part, helper_joint_rand_part) = match agg_id {
+                0 => (&own_joint_rand_part, &other_joint_rand_part),
+                1 => (&other_joint_rand_part, &own_joint_rand_part),
+                _ => return Err(SzkError::Query("bad agg ID".to_string())),
+            };
+            let (joint_rand_seed, joint_rand) =
+                self.derive_joint_rands(leader_joint_rand_part, helper_joint_rand_part);
+            (joint_rand, Some([joint_rand_seed, own_joint_rand_part]))
+        } else {
+            (Vec::new(), None)
+        };
+        let verifier_share = self.typ.query(
+            input_share,
+            proof_share.as_ref(),
+            &query_rand,
+            &joint_rand,
+            2,
+        )?;
+        Ok(SzkQueryShares::new(jr_seed_and_part, verifier_share))
+    }
+
+    /// Returns true if the verifier message indicates that the input from which it was generated is valid.
+    fn decide(
+        &self,
+        verifier: &[T::Field],
+        l_jr_part: Option<Seed<SEED_SIZE>>,
+        h_jr_part: Option<Seed<SEED_SIZE>>,
+        joint_rand_seed_opt: Option<Seed<SEED_SIZE>>,
+    ) -> Result<bool, FlpError> {
+        // Check if underlying FLP proof validates
+        let check_flp_proof = self.typ.decide(verifier)?;
+        if !check_flp_proof {
+            return Ok(false);
+        }
+        // Check that joint randomness was properly derived from both aggregators' parts
+        if self.typ.joint_rand_len() > 0 {
+            let check_joint_rand_seed =
+                self.derive_joint_rand_seed(&l_jr_part.unwrap(), &h_jr_part.unwrap());
+            let joint_rand_seed = joint_rand_seed_opt.unwrap();
+            if check_joint_rand_seed != joint_rand_seed {
+                return Ok(false);
+            };
+        };
+
+        Ok(true)
+    }
+}
+
+mod tests {
+    use super::*;
+    use crate::field::Field128 as TestField;
+    use crate::flp::types::Sum;
+    use rand::{thread_rng, Rng};
+
+    #[test]
+    fn test_sum() {
```

Comment:
I'd like to see a little bit more test coverage.
1. Let's test a `Type` that uses joint randomness and another type that doesn't. That way we ensure all code paths are covered.
2. We should have some negative tests as well, e.g., tweaking: (a) an input share; (b) a proof share; (c) a query share; (d) or the expected joint randomness seed.

To make this easier, it's a good idea to define here a "tester" function that is generic in the `Type`. Then you'd have a `#[test]` for each variant you want to test.